File: java/MiniChatServer.java
import java.io.IOException;
import java.io.InputStream;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;

public class MiniChatServer {

    private static final int port = 6666;

    private static List<Socket> clientSockets = new ArrayList<>();

    private static void initServer() throws IOException {
        System.out.println("init mini_chat server....");
        ServerSocket serverSocket = new ServerSocket(port);

        while (true) {
            Socket socket = serverSocket.accept();
            new Thread(new Runnable() {
                @Override
                public void run() {
                    transfer(socket);
                }
            }).start();
            clientSockets.add(socket);
        }
    }

    private static void transfer(Socket currentSocket) {
        while (true) {
            try {
                byte[] bytes = new byte[1024];
                InputStream inputStream = currentSocket.getInputStream();
                inputStream.read(bytes);
                for (Socket socket : clientSockets) {
                    if (socket == currentSocket) {
                        continue;
                    }
                    System.out.println("recv message from client:" + socket.getRemoteSocketAddress().toString() + " message:" + new String(bytes, 0, bytes.length, "UTF-8"));
                    socket.getOutputStream().write(bytes);
                }
            } catch (Exception e) {
                throw new RuntimeException();
            }

        }

    }

    public static void main(String[] args) {

        try {
            initServer();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: python/mini_chat_server.py
import socket
import thread

cli_fds = []


def init_socket_server():
    ip = "127.0.0.1"
    port = 6666
    server_socket = socket.socket()

    server_socket.bind((ip, port))

    server_socket.listen(100)

    while True:
        cli_fd, add = server_socket.accept()

        print("welcome client:" + str(cli_fd.getsockname()))

        cli_fds.append(cli_fd)

        thread.start_new_thread(recv_message, (cli_fd,))


def recv_message(cli_fd):
    if None == cli_fd:
        ex = Exception("current_fd is NULL")
        raise ex

    while True:
        msg = cli_fd.recv(1024)
        print("recv from client data:%s\n", msg)
        if len(msg) == 0:
            continue

        for fd in cli_fds:
            if fd == cli_fd:
                continue
            fd.send(msg)



if __name__ == '__main__':
    print("mini_chat server init ...")
    init_socket_server()
